Deflake TestAgentForward

zmb3 opened the pull request:
This was another case of the t.TempDir() being cleaned up while the
audit logger is still writing to the directory, which happens when
tests don't properly clean up after themselves. Ensure that any
services we spin up closed via cleanup actions.

Updates #9492
Fixes #5333

zmb3 committed:
Attempt to deflake TestAgentForward

This was another case of the t.TempDir() being cleaned up while the
audit logger is still writing to the directory, which happens when
tests don't properly clean up after themselves. Ensure that any
services we spin up closed via cleanup actions.

Joerger commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -678,14 +682,12 @@ func TestAgentForward(t *testing.T) {
+		return err != nil
+	},
+		150*time.Millisecond, 50*time.Millisecond,
+		"expected socket to be closed, still coud dial")
```
```suggestion
		"expected socket to be closed, still could dial")
```

zmb3 commented:
@Joerger and @r0mant I think I've fixed this for good now. See latest commit.

zmb3 commented:
Well, I've improved things and changed the error.

It used to reproduce after 2-3 minutes with the tempdir cleanup error.

Now it takes 10+ minutes but occasionally fails with a different error:
```
--- FAIL: TestAgentForward (3.04s)
    sshserver_test.go:625: 
                Error Trace:    sshserver_test.go:625
                Error:          Received unexpected error:
                                EOF
                Test:           TestAgentForward
FAIL
FAIL    github.com/gravitational/teleport/lib/srv/regular       3.711s
FAIL
```

zmb3 commented:
Interesting thing is the godocs say:

```
// RequestAgentForwarding sets up agent forwarding for the session.
// ForwardToAgent or ForwardToRemote should be called to route
// the authentication requests.
```

Yet nowhere in the tests do we call `ForwardToAgent` or `ForwardToRemote`.
I"m not familiar enough with this code to know if that's a problem.

zmb3 committed:
Prevent the use of disk-based logging in TestAgentForward

The disk-based logger runs a background process to complete uploads,
which occaisionally fails to finish before the test cleanup tries
to remove the temporary directory.

There are two ways to prevent the use of a disk based logger:

1. Set IsTestStub on the SSH *ServerContext
2. Use a sync session recording mode

Option 2 was selected, because the ServerContext is created by the SSH
server instead of the test, so plumbing that value through would be
a larger change, and I generally dislike test specific modes that can
be mistakenly enabled in non-test situations.

Additionally, update the lib/srv/regular test fixture to allow for
configuring the audit log to use. This allows us to set up a dicarding
logger, since these tests are about agent forwarding behavior and not
audit logging.

zmb3 commented:
This fixes the flakiness, but I noticed a separate issue which I've filed here: https://github.com/gravitational/teleport/issues/13335

zmb3 committed:
Merge branch 'master' into zmb3/deflake-test-agent-forward

zmb3 committed:
Merge branch 'master' into zmb3/deflake-test-agent-forward
